botocore/signers.py

```diff
@@ -411,29 +436,48 @@ def generate_presigned_post(self, request_dict, fields=None,
             policy['conditions'].append(condition)
 
         # Obtain the appropriate signer.
-        query_prefix = '-presign-post'
+        query_suffix = '-presign-post'
         signature_version = self._request_signer.signature_version
-        if not signature_version.endswith(query_prefix):
-            signature_version += query_prefix
-
-        kwargs = {'signing_name': self._request_signer.signing_name,
-                  'region_name': region_name,
-                  'signature_version': signature_version}
+        if signature_version is not botocore.UNSIGNED and not \
+                signature_version.endswith(query_suffix):
+            signature_version += query_suffix
+
+        event_emitter = self._request_signer.event_emitter
+        handler, response = event_emitter.emit_until_response(
+            'choose-signer.s3.PutObject',
+            signing_name=self._request_signer.signing_name,
+            region_name=region_name,
+            signature_version=signature_version)
 
-        signature_type = signature_version.split('-', 1)[0]
+        if response is not None:
+            signature_version = response
 
-        try:
-            auth = self._request_signer.get_auth_instance(**kwargs)
-        except UnknownSignatureVersionError:
-            raise UnsupportedSignatureVersionError(
-                signature_version=signature_type)
+        if signature_version is botocore.UNSIGNED:
```

Comment:
I feel like this a lot of just copying and pasting from generate_presigned_url. Can we consolidate this as well into one helper method?